chg(meta/main): Fixing linting issue

Spreadcat opened the pull request:
##### Summary

Ansible-lint and yamllint always complain about the missing role name and namespace definitition in `./meta/main.yml`.
Also the min_ansible_version is interpreted as a float number instead of a string - another complaint.
other changes are a consistent order of:

1. documentation, then
2. parameter

The comments are now intended as the text/parameters they belong to.
Since `./meta/main.yml` is a yaml file, the file now starts with `---` as well.

Fixes #84150

##### ISSUE TYPE

- Feature Pull Request

##### ADDITIONAL INFORMATION

If you run `ansible-galaxy role init example.test` , then enter the role and run `yamllint .` and `ansible-lint .`, the 
output will cover a number of violations regarding

* comment indentation
* missing yaml document headers
* minimum ansible version
* ~~missing field `namespace`~~
* ~~missing field `role_name`~~

in addition (but ignored in this PR)

* Author name should be changed
* License should be chosen.
* Company should be changed

This PR is trying to fix at least the easy fixable violations and reduce the starting work when creating a new role.

Commit:
chg(meta/main): Fixing linting issue

Ansible-lint always complains about the missing role name and namespace
definitition in `./meta/main.yml`.
Also the min_ansible_version is interpreted as a float number instead of
a string - another complaint.
other changes are a consistent order of:

1. documentation
2. parameter

Also the comments are now intended as the text.
Since `./meta/main.yml` is a yaml file, the file now starts with `---`
as well.

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1083,6 +1083,7 @@ def execute_init(self):
                 company='your company (optional)',
                 license='license (GPL-2.0-or-later, MIT, etc)',
                 role_name=obj_name,
+                namespace=obj_name.split('.')[0],
```
You can supplement/override these variables in 2.17 by using `--extra-vars`. I don't think this is a good default variable to add. If there isn't a '.' in the role name, the role_name and namespace variables will be the same value. When I import a role with `ansible-galaxy role import s-hertel role`, the github user is the correct role namespace, but `ansible-galaxy role init role` doesn't have this info.

Spreadcat replied:
Actually I am a bit unclear on what a role name should be: 

* namespace.role_name
* role_name

`ansible-lint` outputs this:

> ERROR    Computed fully qualified role name of testdir does not follow current galaxy requirements.

And just refers to using the format `namespace.rolename instead. But I struggle to find that information even in the [Ansible Galaxy documentation](https://docs.ansible.com/ansible/latest/galaxy/user_guide.html). The command `ansible-galaxy` certainly takes both formats and creates the role structure in the file system.

My gutt-feeling tells me that this is either unspecified or left this flexible on purpose and probably even is a different discussion than this one.

> You can supplement/override these variables in 2.17 by using --extra-vars

I tried `ansible-galaxy role init --init-path /tmp testdir --extra-vars namespace=spreadcat` on 2.17.4, but namespace didn't show up in `meta/main.yml`. Did I misunderstand you or am I using it wrong?

s-hertel replied:
You probably need to search the ansible-lint documentation, instead of the Galaxy user guide. ansible-lint is more opinionated than the actual feature.

> I tried ansible-galaxy role init --init-path /tmp testdir --extra-vars namespace=spreadcat on 2.17.4, but namespace didn't show up in meta/main.yml. Did I misunderstand you or am I using it wrong?

You would have to provide a role skeleton https://docs.ansible.com/ansible/latest/reference_appendices/config.html#galaxy-role-skeleton that actually uses the variables you pass in. A role skeleton looks like a role, except any .j2 files outside of the templates/ dir will be templated during `ansible-galaxy role init`. Correct usage could look like `ansible-galaxy role init --role-skeleton /path/to/skeleton --extra-vars @/path/to/vars_file.yml rolename`

s-hertel commented on lib/ansible/galaxy/data/default/role/meta/main.yml.j2:
```diff
@@ -1,8 +1,11 @@
   description: {{ description }}
   company: {{ company }}
+  role_name: {{ role_name.split('.')[1] }}
+  namespace: {{ namespace }}
```
The role_name line is causing the test failures. `ansible-galaxy` doesn't expect the role name will be in the format namespace.name. `role_name` and `namespace` are optional Galaxy features, and setting them as part of the default template breaks using the Galaxy-provided defaults. If this is necessary for your project, I think it would be better suited in a custom `--role-skeleton` that uses `--extra-vars`.

Spreadcat replied:
After a bit more of thinking, I come to agree. I forgot about the role skeletons completely.

s-hertel commented on lib/ansible/galaxy/data/default/role/meta/main.yml.j2:
```diff
@@ -17,22 +20,23 @@ galaxy_info:
+  # If this a Container Enabled role, provide the minimum Ansible Container
+  # version.
   # min_ansible_container_version:
+  min_ansible_version: "{{ min_ansible_version }}"
```
The linter rule is wrong, but you can ensure it's a YAML string by including a patch version (doesn't need to be quoted then): `ansible-galaxy role init role --extra-vars min_ansible_version=2.18.0`.

Spreadcat replied:
Yes, that seems to be slightly more elegant, I agree.

> The linter rule is wrong, ...

How do you mean that? I am trying to find some guidelines on entries like that, but they are sparse.


The change now should extract the current Ansible version and add it to the variable.
I was thinking first, that 2.1 might be a sane default, but I fail to see that. The previous code just simply cut after the first three characters, så _2.19_ became _2.1_ .. so did _2.18_.

s-hertel replied:
> How do you mean that?

It's a restriction `ansible-lint` has created rather than reflecting a guideline/limitation from Galaxy, which is the primary user of the `galaxy_info` dictionary.

> I was thinking first, that 2.1 might be a sane default, but I fail to see that. The previous code just simply cut after the first three characters, så 2.19 became 2.1 .. so did 2.18.

Rather than the regex, this would be a tidier way to get full the major.minor: 

```python
            semantic_version = SemanticVersion.from_loose_version(LooseVersion(ansible_version))
            min_ansible_version = f"{semantic_version.major}.{semantic_version.minor}"
```

And personally, I was in favor of a similar code change, but it was ultimately rejected by the team https://github.com/ansible/ansible/pull/78081. (That was opened before --extra-vars support was added though, so the situation is a little more flexible now.)

Spreadcat replied:
I agree, this is more elegant.

s-hertel commented on lib/ansible/galaxy/data/default/role/meta/main.yml.j2:
```diff
@@ -17,22 +20,23 @@ galaxy_info:
 
+  # List tags for your role here, one per line. A tag is a keyword that
+  # describes and categorizes the role. Users find roles by searching for tags.
+  # Be sure to remove the '[]' above, if you add tags to this list.
```
These comments provided a visual representation where to add a multi-line list. The first/last line makes no sense now, same for the dependencies comment below. Also see https://github.com/ansible/ansible/pull/82529.

Spreadcat replied:
I actually needed to read your comment and the comment in the file twice in order to figure out what was wrong. I've changed the wording slightly to adjust. It should be better now.

Spreadcat committed:
chg: min_ansible_version to string

Changing the string definition of min_ansible_version in
`./meta/main.yml`. Removing the hyphens will versions like 2.1 be
interpreted as a float, but the regex should match the whole version
string with three numbers (2.1.0), which then becomes a strint.

Spreadcat committed:
fix: comment

Changing the wording in a comment to reference the correct change that
needs to be done when editing the file.

Spreadcat committed:
chg: disableing role_name and namespace

this should make it easier to focus on the other issue.

Spreadcat committed:
fix: rm role_name and namespace from template

I forgot, I forgot.

Spreadcat committed:
fix(regex): Inavalid escape sequence \d

Spreadcat committed:
fix(regex): Using raw pattern instead.

Spreadcat committed:
rm: namespace comment

Unnecessary. Can be solved by role-skeletons

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1088,7 +1088,7 @@ def execute_init(self):
                 repository_url='http://example.com/repository',
                 documentation_url='http://docs.example.com',
                 homepage_url='http://example.com',
-                min_ansible_version=ansible_version[:3],  # x.y
```
Is `ansible_version` always a string?

Spreadcat replied:
I don't know since I haven't checked the code for that. But IIRC is the particular line with `[:3]` originating from the current devel state of the repository and not a change I made. But I might have misinterpreted your question.

s-hertel commented:
@Spreadcat Thanks for your interest in Ansible, and sorry to review before this was discussed this by the team. We have decided not to include this due to the reasons I've listed in https://github.com/ansible/ansible/issues/84150#issuecomment-2432671476.
